### DISCUSSION TITLE
How to feed in phonemes as input to the released models

### DISCUSSION THREAD

**Discussion by Mu-Y** [coqui-ai/TTS#503] [General]
I'm interested in directly feeding a phoneme sequence (English) to the released models to synthesize audio. From the use cases shown [here](https://github.com/coqui-ai/TTS#example-synthesizing-speech-on-terminal-using-the-released-models), it seems like the input has to be raw text but I've seen `use_phonemes` in the config files. I'm wondering whether any of the released models support taking phonemes as input? Which model should I use? And is it possible to use the `tts` command as [here](https://github.com/coqui-ai/TTS#example-synthesizing-speech-on-terminal-using-the-released-models) for this purpose? Any guidance about running this is appreciated. 

BTW, i read the release notes and it looks like v0.0.14 does not support taking phonemes as input, am I right? So I should only consider the released models up to v0.0.13? 

Thanks a lot!

**Comment by mbarnig** [comment]
I am also interested in this working. I did some tests with versions 0.0.13 and 0.0.14.
I used the following scripts without a configuration file 
```   
tts --out_path /workspace/myTTS-Project/audio/test1.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --text "How are you?"
tts --out_path /workspace/myTTS-Project/audio/test1a.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --text "h ˈaʊ ˈɑ ɹ j ˈu ‖"
```
and the next scripts with two different configuration files where I set the parameter `"use_phonemes": true` (config.json) or `"use_phonemes": false` (config2.json) :     
```    
tts --out_path /workspace/myTTS-Project/audio/test2.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --config_path /workspace/myTTS-Project/config.json --text "How are you?"  
tts --out_path /workspace/myTTS-Project/audio/test2a.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --config_path /workspace/myTTS-Project/config.json --text "h ˈaʊ ˈɑ ɹ j ˈu ‖" 
tts --out_path /workspace/myTTS-Project/audio/test3.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --config_path /workspace/myTTS-Project/config2.json --text "How are you?"  
tts --out_path /workspace/myTTS-Project/audio/test3a.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --config_path /workspace/myTTS-Project/config2.json --text "h ˈaʊ ˈɑ ɹ j ˈu ‖"  
```   
In all six cases a wav file is generated, but with different sizes : 60 KB, 122,4 KB, 61 KB, 117,8 KB, 61,5 KB, 132,7 KB.   
The audio-files with character input sound identical, but have different audio-signals, the audio-files with phoneme input are not intelligible, but seem to pronounce the phonemes :   
![test1](https://user-images.githubusercontent.com/1360633/119259979-07e5ed00-bbd1-11eb-8fb9-504e0d03ba12.jpg)
![test2-3](https://user-images.githubusercontent.com/1360633/119259989-0fa59180-bbd1-11eb-8599-3d540ec21fe0.jpg)

In a last trial I replaced the characters with phonemes in the configuration file :   
```
{
    "model": "Tacotron2-DDC",
    "run_name": "inference-test",
    "run_description": "Tacotron2 with dynamic_convolution and gradual training.",   
    "use_phonemes": false, 
     // VOCABULARY PARAMETERS
    "characters":{
    "characters": "iyɨʉɯuɪʏʊeøɘəɵɤoɛœɜɞʌɔæɐaɶɑɒᵻʘɓǀɗǃʄǂɠǁʛpbtdʈɖcɟkɡqɢʔɴŋɲɳnɱmʙrʀⱱɾɽɸβfvθðszʃʒʂʐçʝxɣχʁħʕhɦɬɮʋɹɻjɰlɭʎʟˈˌːˑʍwɥʜʢʡɕʑɺɧɚ˞ɫ"
    },
}
```
The result was not very different :   
![test4](https://user-images.githubusercontent.com/1360633/119259995-192ef980-bbd1-11eb-93a2-8823747c7c1e.png)

How can we create intelligible output with phoneme input in version => 0.0.14 ? Thank you for your advice.

**Reply by mbarnig** [reply]
To complete my tests I used the model `tts_models/de/thorsten/tacotron2-DCA` with version 0.0.14. As described in the release notes this model is not supported in this new Coqui-TTS version. A runtime error
```    
  File "/TTS/TTS/utils/manage.py", line 106, in download_model
    raise RuntimeError(
RuntimeError:  [!] Use 🐸TTS <= v0.0.13 for this model. Current version does not support phoneme based models.
root@58859999859d:/TTS/TTS# 
```   
is thrown in all three test cases (without config-file, with config-file `"use_phonemes": true`, with config-file `"use_phonemes": false`. 

My hope was that one configuration is transparent to an inference input with phonemes.

**Comment by lexkoro** [comment]
If I'm not mistaken, you can't currently enter phonemes directly. The `use_phonemes` defines that the input "should" be converted to phonemes, but not that you can enter phonemes directly.

Here is where the text input is being converted to the input sequence for the network
https://github.com/coqui-ai/TTS/blob/5482a0f62d3da9821b4689f4e9268580904a081f/TTS/tts/utils/synthesis.py#L265

Following this you can check how the input of graphemes is handled and how the output looks like.
https://github.com/coqui-ai/TTS/blob/5482a0f62d3da9821b4689f4e9268580904a081f/TTS/tts/utils/text/__init__.py#L63

I think you could write a function that just processes phonemes when they are entered directly.

**Reply by Mu-Y** [reply]
Hello, 
thanks for your reply. If I understand correctly, by "phonemes", it means that all released models are using IPA phonemes, right? Is there a way to use, e.g. Arpabet phonemes (http://www.speech.cs.cmu.edu/cgi-bin/cmudict)?  Has anyone done this before and how does the performance look like?

**Reply by mbarnig** [reply]
Hello @Mu-Y. My understanding is that you can specify the phonemes in the configuration file. You probably can replace the IPA phonemes by [ARPABET](https://en.wikipedia.org/wiki/ARPABET) to train an new english model, but I don't see the advantage to do it.

**Comment by mbarnig** [comment]
Thank you @SanjaESC for your hints where to start exploring the code. It's not easy to understand the whole workflow, but my feeling after looking at the code was that it's possible to enter phonemes for inference. I did some additional tests. When entering the phonemes without stress and adding an eos-symbol at the end, the generated wav file is OK.

Here are the results from my terminal and the waveforms in Audacity for inference with characters and with phonemes of the sentence "How are you?", without specifying a configuration file, using Coqui-TTS version 0.0.14.
```   
root@58859999859d:/TTS# 
tts --out_path /workspace/myTTS-Project/audio/test-chars.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --text "How are you?"
 > tts_models/en/ljspeech/tacotron2-DDC is already downloaded.
 > vocoder_models/en/ljspeech/hifigan_v2 is already downloaded.
 > Using model: Tacotron2
 > Generator Model: hifigan_generator
Removing weight norm...
 > Text: How are you?
 > Text splitted to sentences.
['How are you?']
 > Processing time: 0.319094181060791
 > Real-time factor: 0.23478466005040183
 > Saving output to /workspace/myTTS-Project/audio/test-chars.wav
 
root@58859999859d:/TTS# 
tts --out_path /workspace/myTTS-Project/audio/test-phons.wav --model_name tts_models/en/ljspeech/tacotron2-DDC --text "h aʊ ɑ ɹ j u ~"
 > tts_models/en/ljspeech/tacotron2-DDC is already downloaded.
 > vocoder_models/en/ljspeech/hifigan_v2 is already downloaded.
 > Using model: Tacotron2
 > Generator Model: hifigan_generator
Removing weight norm...
 > Text: h aʊ ɑ ɹ j u ~
 > Text splitted to sentences.
['h aʊ ɑ ɹ j u ~']
 > Processing time: 0.366788387298584
 > Real-time factor: 0.24290256907537772
 > Saving output to /workspace/myTTS-Project/audio/test-phons.wav 
 ```    

![test-chars-phons](https://user-images.githubusercontent.com/1360633/119336164-9324b880-bc8d-11eb-8818-c4fa8abd0671.png)

 I will continue my tests in Coqui-TTS version 0.0.13 with the german model and report about my findings.

**Comment by Mu-Y** [comment]
As a quick workaround, I made a tiny modification to the `phoneme_to_sequence` function in `TTS/tts/utils/text/__init__.py`: I commented line 71 and 72, and added 
```
to_phonemes=text
```
below line 72.

Now, from the `tts` CLI, we can directly play with the IPAs:

```
~/TTS$ tts --text="ð|ɛɹ| |w|ʌ|z| |n|ʌ|θ|ɪ|ŋ| |f|ɔː|ɹ| |ʌ|s| |b|ʌ|t| |ð|ə| |w|aɪ|d| |ɹ|ɔː
| |oʊ|ʃ|ə|n|" --model_name="tts_models/en/ljspeech/tacotron2-DCA" --out_path=output/test_phones2.wav --use_cuda
=True                                                                                                          
 > tts_models/en/ljspeech/tacotron2-DCA is already downloaded.                                                 
 > vocoder_models/en/ljspeech/multiband-melgan is already downloaded.                                          
 > Using model: Tacotron2                                                                                      
 > Generator Model: multiband_melgan_generator                                                                 
 > Text: ð|ɛɹ| |w|ʌ|z| |n|ʌ|θ|ɪ|ŋ| |f|ɔː|ɹ| |ʌ|s| |b|ʌ|t| |ð|ə| |w|aɪ|d| |ɹ|ɔː| |oʊ|ʃ|ə|n|                     
 > Text splitted to sentences.                                                                                 
['ð|ɛɹ| |w|ʌ|z| |n|ʌ|θ|ɪ|ŋ| |f|ɔː|ɹ| |ʌ|s| |b|ʌ|t| |ð|ə| |w|aɪ|d| |ɹ|ɔː| |oʊ|ʃ|ə|n|']                          
 > Processing time: 0.16599321365356445                                                                        
 > Real-time factor: 0.05476643465796469                                                                       
 > Saving output to output/test_phones2.wav
```

Audio sounds reasonable. (I'm using v0.0.13). 
Note that in this way the input IPA has to be under a specific pattern: `|` and `whitespace` have to be used properly like in this example (`|` within a word, `whitespace` between two words).

**Reply by mbarnig** [reply]
@Mu-Y. Great. That's what I wanted to do. Thank you very much for this elegant modification. I will give it a try today and report about my results.

**Comment by saona-raimundo** [comment]
I am also interested in this feature! Is this going to be available through the CLI at some point?

**Reply by erogol** [reply]
We don't have it in our current plans. However, you can drop a future request if you want.

**Reply by saona-raimundo** [reply]
Great! Just to be sure, would the phonemes-to-speech be a byproduct of training a tts model?

If yes, just looking into the future, do you think it is worth using such byproduct instead of training from the ground up a phonemes-to-speech model (the data is directly phonemes instead of texts).

Thanks beforehand!

**Comment by swesterfeld** [comment]
> As a quick workaround, I made a tiny modification to the `phoneme_to_sequence` function in `TTS/tts/utils/text/__init__.py`: I commented line 71 and 72, and added
> 
> ```
> to_phonemes=text
> ```
> 
> below line 72.

I'm trying to reproduce being able to use phonemes as input directly, however the function you mentioned is no longer present in the file you mentioned. Is there a way to do it without changing the source code with the current TTS version? Or do I have to change some other function these days?

Also, I've seen that https://coqui.ai/ supports editing the phoneme sequence directly so that I can control
- pitch
- energy
- duration

of each phoneme individually. I'd like to do that too without using their non-free online-only editor. Ideally I'd like to use something similar to mbrola .pho files (https://github.com/numediart/MBROLA/blob/master/README.md#phoneme-commands) to be able to control timing and pitch directly.
